### PR TITLE
get rid of existsAs in (some) tracking code

### PR DESCRIPTION
**Pull request by mtosi**
this PR should be almost identical to https://github.com/cms-sw/cmssw/pull/12966
which can be close, indeed

I see that it cannot be automatically merge :(
how could I figure it out which is the source of the problem ?
thanks

### CODE REVIEW

**Comment by slava77** [comment]
@mtosi 
my guess is that the conflicts are with #13037
you can either wait for the next IB
or rebase your work to CMSSW_8_0_X

**Comment by Martin-Grunewald** [comment]
@mtosi 
If this one replaces #12966 then YOU should actually close that PR!

**Comment by Martin-Grunewald** [comment]
@mtosi 

Please add a comment in the file ```customizeHLTfor2016trackingTemplate```
and tell what it is used for and what it is doing in somewhat more detail.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -44,6 +44,40 @@ def customiseFor12387(process):
             setattr(producer,'LegacyUnpacker',cms.bool(False))
     return process
 
+
+def customiseFor12966(process):
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: PR number now has changed (why didn't you UPDATE the old PR????)

**Comment by mtosi** [comment]
fakerate_hltiter2merged
effic_hltiter2merged

**Comment by Martin-Grunewald** [comment]
I add the e-mail exchange here in the github PR to have a record.
Andrea Bocci's questions and their answers:
```
> 1. if #13062 contains all the changes from #13055, then close #13055
>
> 2. is customiseFor2016trackingTemplate() supposed to be called after the
> customisation in customizeHLTforCMSSW, or instead of it ? The former
should
> be OK, the latter would be a problem.
>
>
#13055 is merged, now

in order to have the tracking template in place
I added the following lines on top of the menu available in the release

# standard customization (which includes PR #13062 from tracking@HLT)
from HLTrigger.Configuration.customizeHLTforALL import customizeHLTforAll
process = customizeHLTforAll(process,_customInfo)

# tracking template for 2016
from HLTrigger.Configuration.customizeHLTfor2016trackingTemplate import
customiseFor2016trackingTemplate
process = customiseFor2016trackingTemplate(process)

open('dump.py','w').write(process.dumpPython())

hope it is fine
  mia
```

**Comment by Martin-Grunewald** [comment]
I add the e-mail exchange here in the github PR to have a record.
Martin's questions and their answers:
```
> It is also not clear to me what is needed when, and whether added
> parameters
> need to be able to be changed by POG/PAG testers or could remain "hidden"
> in customisation files until we parse pre6.
>
> needed parameters hidden in the customization file
should be available in the pre6 confDB templates
they should be handled by the fillDescription
and if it is not the case, because they are global PSet,
I have to add them
indeed, I would ask you to ping me when the confDB templates will be
available (thanks)


> It is also not clear whu you have both a modifcation in the std.
> custom file, AND you have added another custom file.

I change the standard one because I want to make sure that the standard
menu (2015) is not affected at all
the std customization is aimed to give you 2015 tracking

the new customization function is meant to provide the 2016 tracking while
the confDB templates and settings
(indeed I have anyhow to propagate the changes in confDB as well)
will be avialble
I did in this way because I understood other POGs and PAGs depend on this
changes and I do not want to increase the delay,
which unfortunately I already generated (even if it was not intended, of
course ... I had silly bugs to deal w/ and I'm still in the turn-on of
having a kid for the first time, I would say)

hope the code is fine and helpful
but, please, let me know if there are still issues/missing points
thanks
  mia
```

File: HLTrigger/Configuration/python/customizeHLTfor2016trackingTemplate.py [hunk 2]
```diff
@@ -0,0 +1,686 @@
...
+            copyExtras       = cms.untracked.bool(False),
+            copyTrajectories = cms.untracked.bool(False)
+          )
+         )
+        )
+        iter0HP  = getattr(process,'hltIter0PFlowTrackSelectionHighPurity')
+
+        iter0seq = getattr(process,'HLTIterativeTrackingIteration0')
+        iter0seq.insert( iter0seq.index( iter0HP ), getattr(process,'hltIter0PFlowTrackCutClassifier') )
+        print 'iter0seq: ', iter0seq
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTfor2016trackingTemplate.py]: Please remove this and all other "print" statements in the py files.
The printouts would end up in the final py file and lead to error as they are not valid python.
In any case the py file should be "silent"

**Comment by mtosi** [comment]
thanks martin for catching my "dirty stuff" !

**Comment by Martin-Grunewald** [comment]
```
> I change the standard one because I want to make sure that the standard
> menu (2015) is not affected at all
> the std customization is aimed to give you 2015 tracking
>

Just to clarify my point of view: we are *not* keeping a 2015-like menu
alive in CMSSW 8.x anyway, so I think that we do not need to keep the 2015
settings as the default, once the 2016 settings are approved.

Once the 2016 menus are done and frozen, it will be a different story - but
right now I think we are free to make any change we want.
```

So the current modification to the std. customisation file is then only
to make the menu work with the new C++ code (and until we parse it into ConfDB).
Likewise for the added file and the 2016 tracking?
So can't we simply go directly to the 2016 tracking in the std. customisation file?

**Comment by fwyzard** [comment]
> So can't we simply go directly to the 2016 tracking in the std.
customisation file?

Yes, that's what I was trying to say on Wednesday...

**Comment by Martin-Grunewald** [comment]
OK, then let's please do so:
remove the "new" custom file, and put the 2016 tk config in the "std" custom file.

**Comment by mtosi** [comment]
added the customiseFor2016trackingTemplate in the standard one

some of the settings in the standard one are anyhow needed because some modules are used by other POGs

File: HLTrigger/Configuration/python/customizeHLTfor2016trackingTemplate.py [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+        iter2seq = getattr(process,'HLTIterativeTrackingIteration2')
+        iter2seq.insert( iter2seq.index( iter2HP ), getattr(process,'hltIter2PFlowTrackCutClassifier') )
+        iter02seq = getattr(process,'HLTIterativeTrackingIter02')
+        if hasattr(iter02seq,'hltIter2PFlowTrackSelectionHighPurity'):
+            iter02seq.insert( iter02seq.index( iter2HP ), getattr(process,'hltIter2PFlowTrackCutClassifier') )
+
+    return process
+
+def customiseFor2016trackingTemplate(process):
+    bug_fix(process)
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTfor2016trackingTemplate.py]: Here and in the following for all calls, you need to write ```process = bug_fix(process)``` etc.

**Comment by slava77** [comment]
merge conflicts ...

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py.py]: Please remove this file as a duplicate with a wrong name anyway: ```customizeHLTforCMSSW.py.py```

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                if not hasattr(pset,'minGoodStripCharge'):
+                    pset.minGoodStripCharge  = cms.PSet(refToPSet_ = cms.string('HLTSiStripClusterChargeCutNone'))
+                if not hasattr(pset,'maxCCCLostHits'):    
+                    pset.maxCCCLostHits      = cms.int32(9999)
+                if not hasattr(pset,'seedExtension'):
+                    pset.seedExtension       = cms.int32(0)
+                if not hasattr(pset,'strictSeedExtension'):
+                    pset.strictSeedExtension = cms.bool(False)
+
+    customiseFor2016trackingTemplate(process)
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py.py]: Also here it must be ```process = customiseFor2016trackingTemplate(process)```.

**Comment by Martin-Grunewald** [comment]
According to github, there are still "merge conflicts"; indeed:
```
Auto-merging HLTrigger/Configuration/python/customizeHLTforCMSSW.py
CONFLICT (content): Merge conflict in HLTrigger/Configuration/python/customizeHLTforCMSSW.py
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by mtosi** [comment]
I thought I fixed the conflicts
I do not understand why it is still complaining ....
I've just re-fixed the file, but still it complains

I really do not understand :(

**Comment by Martin-Grunewald** [comment]
The error is still the same:
```
Auto-merging HLTrigger/Configuration/python/customizeHLTforCMSSW.py
CONFLICT (content): Merge conflict in HLTrigger/Configuration/python/customizeHLTforCMSSW.py
Automatic merge failed; fix conflicts and then commit the result.
Unable to merge branch refs/pull/13062/head from repository cms-sw.
```
Did you follow the suggested remedy?
```
Automatic merge failed; fix conflicts and then commit the result.
```
?

**Comment by Martin-Grunewald** [comment]
@mtosi 
This is getting us nowhere, the PR is still unmergable. I have redone the PR, see: #13164.
Please check(!!) and eventually sign and close this unmergeable one.

**Pull request closed without merging**